parser/protobuf_usp.go

```diff
@@ -0,0 +1,152 @@
+	recordMetric := skogul.Metric{
+		Time:     nil,
+		Metadata: p.createRecordMetadata(record),
+		Data:     data,
+	}
+
+	if recordMetric.Data == nil || recordMetric.Metadata == nil {
+		atomic.AddUint64(&p.stats.NilData, 1)
+		return nil, errors.New("Metric metadata or data was nil; aborting")
+	}
+
+	container := skogul.Container{}
+	container.Metrics = make([]*skogul.Metric, 1)
+	container.Metrics[0] = &recordMetric
+	return &container, err
+}
+
+// Unmarshals []byte into a protoc generated struct and returns it
+func (p *P) getUspRecord(d []byte) (*usp.Record, error) {
+	unmarshaledMessage := &usp.Record{}
+	if err := proto.Unmarshal(d, unmarshaledMessage); err != nil {
+		atomic.AddUint64(&p.stats.ParseErrors, 1)
+		return nil, fmt.Errorf("Failed to unmarshal protocol buffer. Error: %w", err)
+	}
+	return unmarshaledMessage, nil
+}
+
+// Unmarshals []byte consisting of the record payload into
+// a protoc generated struct and returns it
+func (p *P) getRecordMsgPayload(payload []byte) (*usp.Msg, error) {
+	msgPayload := &usp.Msg{}
+
+	if err := proto.Unmarshal(payload, msgPayload); err != nil {
+		atomic.AddUint64(&p.stats.ParseErrors, 1)
+		return nil, fmt.Errorf("Failed to unmarshal payload. Error: %w", err)
+	}
+
+	return msgPayload, nil
+}
+
+// Creates a map[string]interface{} of the metadata for skogul.Metric
+func (p *P) createRecordMetadata(h *usp.Record) map[string]interface{} {
+	var d = make(map[string]interface{})
+
+	d["from_id"] = h.GetFromId()
+	d["to_id"] = h.GetToId()
+	d["payload_security"] = h.GetPayloadSecurity()
+	d["sender_cert"] = h.GetSenderCert()
+	d["version"] = h.GetVersion()
+	d["mac"] = h.GetMacSignature()
+	return d
+}
+
+// Unmarshals event parameters to json
+func (p *P) extractJSON(s string) (map[string]interface{}, error) {
+	input := []byte(s)
+
+	var d map[string]interface{}
+
+	if err := json.Unmarshal(input, &d); err != nil {
+		return nil, err
+	}
+
+	return d, nil
+}
+
+// Creates a map[string]interface{} of the record payload for skogul.Metric
+func (p *P) createRecordData(t *usp.Record) (map[string]interface{}, error) {
+	var jsonMap = make(map[string]interface{})
+	payload, err := p.getRecordMsgPayload(t.GetNoSessionContext().GetPayload())
+
+	if err != nil {
+		return nil, err
+	}
+
+	// Check if request contains the Notify event. (It could be a different event by mistake)
+	if d, ok := payload.Body.GetRequest().GetReqType().(*usp.Request_Notify); !ok {
+		return nil, fmt.Errorf("Invalid event %s", d.Notify.GetEvent())
+	}
+
+	jsonData, err := p.extractJSON(payload.GetBody().GetRequest().GetNotify().GetEvent().GetParams()["Data"])
+
+	if err != nil {
+		return nil, err
+	}
+
+	jsonMap["event"] = payload.GetBody().GetRequest().GetNotify().GetEvent().GetObjPath()
+	jsonMap["event_type"] = payload.GetBody().GetRequest().GetNotify().GetEvent().GetEventName()
+	jsonMap["subscription_id"] = payload.GetBody().GetRequest().GetNotify().GetSubscriptionId()
+	jsonMap["event_parameters"] = jsonData["Report"]
```

Comment:
I think some of this might be considered metadata? Except jsonData["Report"]? Let's take a look at some real-life data to evaluate before deciding.